stock_picking_batch_creation/wizards/make_picking_batch.py

```diff
@@ -0,0 +1,349 @@
+        # - maximum number of preparation lines is not outreached
+        # - numbers of bins available is greater than 0
+        # - The device for the current picking is supposed to be
+        # the same as the one for the first picking
+        total_weight = first_picking.total_weight_batch_picking
+        total_volume = first_picking.total_volume_batch_picking
+        total_nbr_picking_lines = first_picking.nbr_picking_lines
+        for picking in pickings[1:]:
+            if max_pickings and len(selected_pickings) == max_pickings:
+                # selected enough!
+                break
+
+            # Check that the picking can go on the selected device
+            picking_device = self._compute_device_to_use(picking)
+            different_device_for_current_picking = device != picking_device
+
+            picking.total_weight_batch_picking = self._picking_weight(picking)
+            picking.total_volume_batch_picking = self._picking_volume(picking)
+            picking.nbr_picking_lines = self._picking_nbr_lines(picking)
+            picking.nbr_bins_batch_picking = math.ceil(
+                picking.total_volume_batch_picking / volume_per_bin
+            )
+
+            weight_limit_outreached = device.max_weight and gt(
+                total_weight + picking.total_weight_batch_picking,
+                device.max_weight,
+                precision_weight,
+            )
+            volume_limit_outreached = device.max_volume and gt(
+                total_volume + picking.total_volume_batch_picking,
+                device.max_volume,
+                precision_volume,
+            )
+            nbr_lines_limit_outreached = (
+                self.maximum_number_of_preparation_lines
+                and gt(
+                    total_nbr_picking_lines + picking.nbr_picking_lines,
+                    self.maximum_number_of_preparation_lines,
+                    1,
+                )
+            )
+            available_bins_outreached = (
+                available_nbr_bins - picking.nbr_bins_batch_picking < 0
+            )
+
+            if not (
+                weight_limit_outreached
+                or volume_limit_outreached
+                or nbr_lines_limit_outreached
+                or available_bins_outreached
+                or different_device_for_current_picking
+            ):
+                # All conditions are OK : we keep the picking in the cluster
+                selected_pickings |= picking
+                total_weight += picking.total_weight_batch_picking
+                total_volume += picking.total_volume_batch_picking
+                total_nbr_picking_lines += picking.nbr_picking_lines
+                available_nbr_bins -= picking.nbr_bins_batch_picking
+            else:
+                # At least one condition KO : we discard the picking from the cluster
+                unselected_pickings |= picking
+
+        return selected_pickings, unselected_pickings
```

Comment:
We should not have a specific code for the first picking and you should exclude a picking int the loop as of a condition is not met.

```suggestion
        selected_picking_ids = []
        total_weight = 0.0
        total_volume = 0.0
        total_nbr_picking_lines = 0
        precision_weight = self._precision_weight()
        precision_volume = self._precision_volume()

        def gt(value1, value2, digits):
            """Return True if value1 is greater than value2"""
            return tools.float_compare(value1, value2, precision_digits=digits) == 1

        # Always take the first picking even if it doesn't fullfill the
        # weight/volume criteria. This allows to always process at least
        # one picking which won't be processed otherwise
        available_nbr_bins = device.nbr_bins
        
        # Add pickings to the batch as long as
        # - total volume is not outreached
        # - total weight is not outreached
        # - maximum number of preparation lines is not outreached
        # - numbers of bins available is greater than 0
        # - The device for the current picking is supposed to be
        for picking in pickings:
            available_nbr_bins == 0:
                break
              
            if max_pickings and len(selected_pickings) == max_pickings:
                # selected enough!
                break
            
            # Check that the picking can go on the selected device
            picking_device = self._compute_device_to_use(picking)
            if device != picking_device:
                continue

             available_bins_outreached = (
                available_nbr_bins - picking.nbr_bins_batch_picking < 0
            )
            if available_bins_outreached :
                 continue

            picking.total_weight_batch_picking = self._picking_weight(picking)
            picking.total_volume_batch_picking = self._picking_volume(picking)
            picking.nbr_picking_lines = self._picking_nbr_lines(picking)
            picking.nbr_bins_batch_picking = math.ceil(
                picking.total_volume_batch_picking / volume_per_bin
            )

            weight_limit_outreached = device.max_weight and gt(
                total_weight + picking.total_weight_batch_picking,
                device.max_weight,
                precision_weight,
            )
            if weight_limit_outreached:
                continue 
            
            volume_limit_outreached = device.max_volume and gt(
                total_volume + picking.total_volume_batch_picking,
                device.max_volume,
                precision_volume,
            )
            if volume_limit_outreached:
                continue
            
            nbr_lines_limit_outreached = (
                self.maximum_number_of_preparation_lines
                and gt(
                    total_nbr_picking_lines + picking.nbr_picking_lines,
                    self.maximum_number_of_preparation_lines,
                    1,
                )
            )
            if nbr_lines_limit_outreached:
                continue
            
            # All conditions are OK : we keep the picking in the cluster
            selected_picking_ids.append(picking.id)
            total_weight += picking.total_weight_batch_picking
            total_volume += picking.total_volume_batch_picking
            total_nbr_picking_lines += picking.nbr_picking_lines
            available_nbr_bins -= picking.nbr_bins_batch_picking

        selected_pickings = self.env["stock.picking"].browse(selected_picking_ids)
        unselected_pickings = pickings - selected_pickings
        return selected_pickings, unselected_pickings
```